File: TOOLKIT/hdfeos/samples/WriteDimscaleGrid.c
#include "hdf.h"
#include "HdfEosDef.h"

/*
 * In this example we will (1) open the "GridFile" HDF file, (2) attach to
 * the "Grid1" grid, and (3) set dimension scale for a dimension in a filed.
 */


main()
{

    intn            status, i, j;
    int32           gdfid, GDid1, GDid2, GDid3;
    int32           xdim, ydim;
    int32           ntime;
    int32           nbands;
    int32           data[10]={3,6,9,12,15,18,23,26,29,32};
    int32           bands[10]={12,14,18};
    char            label[16];
    char            unit[16];
    char            format[16];

    /*
     * We first open the HDF grid file, "GridFile.hdf".  Because this file
     * already exist and we wish to write to it, we use the DFACC_RDWR access
     * code in the open statement.  The GDopen routine returns the grid file
     * id, gdfid, which is used to identify the file in subsequent routines.
     */

    gdfid = GDopen("GridFile_created_with_hadeos_sample_file_writer_of_HDFEOS2_version_219_or_higher_release.hdf", DFACC_RDWR);


    /*
     * If the grid file cannot be found, GDopen will return -1 for the file
     * handle (gdfid).  We there check that this is not the case before
     * proceeding with the other routines.
     * 
     * The GDattach routine returns the handle to the existing grid "Grid1",
     * GDid.  If the grid is not found, GDattach returns -1 for the handle.
     */

    if (gdfid != -1)
    {
	/* the field Pollution has Time,YDim,XDim dimensions. 
	   see SetupGrid.c and DefineGDflds.c 
	   xdim = 120;
	   ydim = 200;
	   Time dim is set to 10
	*/

	GDid1 = GDattach(gdfid, "UTMGrid");
	if (GDid1 == -1)
	  {
	    printf("\t\tError: Cannot attach to grid \"UTMGrid\"\n");
	    GDclose(gdfid);
	    return -1;
	  }

	xdim = 120;
	ntime = 10;
	
	status = GDsetdimscale(GDid1, "Pollution", "XDim", xdim, DFNT_FLOAT64, NULL);
	if (status == -1)
	  {
	    printf("\t\tError: Cannot set Dimension Scale for XDim dimemnsion in field \"Pollution\"\n");
	    GDdetach(GDid1);
	    GDclose(gdfid);
	    return -1;
	  }

	status = GDsetdimscale(GDid1, "Pollution", "Time", ntime, DFNT_INT32, data);
	if (status == -1)
	  {
	    printf("\t\tError: Cannot set Dimension Scale for Time dimemnsion in field \"Pollution\"\n");
	    GDdetach(GDid1);
	    GDclose(gdfid);
	    return -1;
	  }

	strcpy(label, "X Dimension");
	strcpy(unit, "meters");
	strcpy(format, "F7.2");

	status = GDsetdimstrs(GDid1, "Pollution", "XDim", label, unit, format);
	if (status == -1)
	  {
	    printf("\t\tError: Cannot set Dimension Scale strs for XDim dimemnsion in field \"Pollution\"\n");
	    GDdetach(GDid1);
	    GDclose(gdfid);
	    return -1;
	  }

	strcpy(label, "Time Dim");
	strcpy(unit, "s");
	strcpy(format, "I3");

	status = GDsetdimstrs(GDid1, "Pollution", "Time", label, unit, format);
	if (status == -1)
	  {
	    printf("\t\tError: Cannot set Dimension Scale strs for Time dimemnsion in field \"Pollution\"\n");
	    GDdetach(GDid1);
	    GDclose(gdfid);
	    return -1;
	  }

	GDdetach(GDid1);

	/* the field Spectra has Bands,YDim,XDim dimensions. 
	   see SetupGrid.c and DefineGDflds.c
	   xdim = 100;
	   ydim = 100;
	   Bands dim is set to 3
	*/
	
	GDid2 = GDattach(gdfid, "PolarGrid");
	if (GDid2 == -1)
	  {
	    printf("\t\tError: Cannot attach to grid \"PolarGrid\"\n");
	    GDclose(gdfid);
	    return -1;
	  }
	
	ydim = 100;
	nbands = 3;
	
	status = GDsetdimscale(GDid2, "Spectra", "YDim", ydim, DFNT_FLOAT64, NULL);
	if (status == -1)
	  {
	    printf("\t\tError: Cannot set Dimension Scale for YDim dimemnsion in field \"Spectra\"\n");
	    GDdetach(GDid2);
	    GDclose(gdfid);
	    return -1;
	  }


	status = GDsetdimscale(GDid2, "Spectra", "Bands", nbands, DFNT_INT32, bands);
	if (status == -1)
	  {
	    printf("\t\tError: Cannot set Dimension Scale for Bands dimemnsion in field \"Spectra\"\n");
	    GDdetach(GDid2);
	    GDclose(gdfid);
	    return -1;
	  }


	strcpy(label, "Y Dimension");
	strcpy(unit, "meters");
	strcpy(format, "F7.2");

	status = GDsetdimstrs(GDid2, "Spectra", "YDim", label, unit, format);
	if (status == -1)
	  {
	    printf("\t\tError: Cannot set Dimension Scale strs for YDim dimemnsion in field \"Spectra\"\n");
	    GDdetach(GDid2);
	    GDclose(gdfid);
	    return -1;
	  }

	strcpy(label, "Bands");
	strcpy(unit, "None");
	strcpy(format, "I2");

	status = GDsetdimstrs(GDid2, "Spectra", "Bands", label, unit, format);
	if (status == -1)
	  {
	    printf("\t\tError: Cannot set Dimension Scale strs for Bands dimemnsion in field \"Spectra\"\n");
	    GDdetach(GDid2);
	    GDclose(gdfid);
	    return -1;
	  }

    }

	GDdetach(GDid2);

    /* the field GeoSpectra has YDim,XDim dimensions. 
       see SetupGrid.c and DefineGDflds.c
       xdim = 100;
       ydim = 100;
       Bands dim is set to 3
    */

    xdim = 60;
    ydim = 40;

    GDid3 = GDattach(gdfid, "GEOGrid");
    if (GDid3 == -1)
      {
	printf("\t\tError: Cannot attach to grid \"GEOGrid\"\n");
	GDclose(gdfid);
	return -1;
      }
    status = GDsetdimscale(GDid3, "GeoSpectra", "YDim", ydim, DFNT_FLOAT64, NULL);
    if (status == -1)
      {
	printf("\t\tError: Cannot set Dimension Scale for YDim dimemnsion in field \"GeoSpectra\"\n");
	GDdetach(GDid3);
	GDclose(gdfid);
	return -1;
      }
    
    strcpy(label, "Y Dimension");
    strcpy(unit, "Decimal Degrees");
    strcpy(format, "F7.2");
    
    status = GDsetdimstrs(GDid3, "GeoSpectra", "YDim", label, unit, format);
    if (status == -1)
      {
	printf("\t\tError: Cannot set Dimension Scale strs for YDim dimemnsion in field \"GeoSpectra\"\n");
	GDdetach(GDid3);
	GDclose(gdfid);
	return -1;
      }
    

    GDdetach(GDid3);
    GDclose(gdfid);

    return 0;
}

File: TOOLKIT/hdfeos/samples/writedimscalegrid.f
!/*
! * In this example we will (1) open the "GridFile" HDF file, (2) attach to
! * the "gdid" grid, and (3) set dimension scale for a dimension in a filed.
! */

	program writedimscalegrid

	integer       j, status, gddetach, gdclose
	integer*4     gdfid, gdid, gdopen, gdattach
	integer*4     gdsetdimscale, gdsetdimstrs
	real*8        veg1(120),veg2(200)
	real*8        temp1(100),temp2(100)
	integer*4     bands(3)
	integer*4     xdim,ydim,nbands

	integer DFACC_RDWR
	parameter (DFACC_RDWR=3)
	integer DFNT_FLOAT32
	parameter (DFNT_FLOAT32=5)
	integer DFNT_FLOAT64
	parameter (DFNT_FLOAT64=6)
	integer DFNT_INT32
	parameter (DFNT_INT32=24)

!    /*
!     * We first open the HDF grid file, "GridFile.hdf".  Because this file
!     * already exist and we wish to write to it, we use the DFACC_RDWR access
!     * code in the open statement.  The gdopen routine returns the grid file
!     * id, gdfid, which is used to identify the file in subsequent routines.
!     */

	gdfid = gdopen("GridFile_created_with_hadeos_sample_file_write"//
     1"r_of_HDFEOS2_version_219_or_higher_release.hdf", DFACC_RDWR)

!    /*
!     * If the grid file cannot be found, gdopen will return -1 for the file
!     * handle (gdfid).  We there check that this is not the case before
!     * proceeding with the other routines.
!     * 
!     * The gdattach routine returns the handle to the existing grid "gdid",
!     * gdid.  If the grid is not found, gdattach returns -1 for the handle.
!     */
	
	if (gdfid .ne. -1) then
	   
	   gdid = gdattach(gdfid, "UTMGrid")
	   
	   if (gdid .ne. -1) then
	      
              xdim = 120;
              ydim = 200;
	      
              status = gdsetdimscale(gdid, "Vegetation", "XDim", xdim,
     1	      DFNT_FLOAT64, veg1)
	      write(*,*) 'status: gdsetdimscale Vegetation:XDim',status
              status = gdsetdimscale(gdid, "Vegetation", "YDim", ydim,
     1	      DFNT_FLOAT64, veg2)
	      write(*,*) 'status: gdsetdimscale Vegetation:YDim',status

!	label, = 'X Dimension'
!	unit, = 'meters'
!	format, = 'F7.2'

              status = gdsetdimstrs(gdid, "Vegetation", "YDim", 
     1	      "Y Dimension", "meters", "F7.2")

	   endif

	   status = gddetach(gdid)

	   gdid = gdattach(gdfid, "PolarGrid")
	   if (gdid .ne. -1) then
	      
	      xdim = 100
	      ydim = 100
	      
	      status = gdsetdimscale(gdid, "Temperature", "XDim", xdim,
     1	      DFNT_FLOAT64, temp1)
	      write(*,*) 'status: gdsetdimscale Temperature:XDim',status
	      write(*,*) 'this -1 was expected for Temperature:XDim'
	      status = gdsetdimscale(gdid, "Temperature", "YDim", ydim,
     1	      DFNT_FLOAT64, temp2)
	      write(*,*) 'status: gdsetdimscale Temperature:YDim',status
	      write(*,*) 'this -1 was expected for Temperature:YDim'
	      nbands = 3;
	      do j=1,nbands
		 bands(j) = j*2+1
	      enddo
	      status = gdsetdimscale(gdid, "Spectra", "YDim", ydim, 
     1	      DFNT_FLOAT64, temp2)
	      write(*,*) 'status: gdsetdimscale Spectra:YDim',status

 	      status = gdsetdimscale(gdid, "Spectra", "Bands", nbands,
     1	      DFNT_INT32, bands);
	      write(*,*) 'status: gdsetdimscale Spectra:Bands',status
	      
	      status = gdsetdimstrs(gdid, "Spectra", "YDim", "YDim Dim",
     1	      "meters", "F7.2")
	      
	      status = gdsetdimstrs(gdid, "Spectra", "Bands", "BANDS",
     1	      "None", "I2")
	      
	      status = gddetach(gdid)
	   endif
	   
	   status = gdclose(gdfid)
	endif
	stop
	end

